### ISSUE TITLE
Asp.net core webapi what package to use?

### ISSUE THREAD

**Issue by WhatFreshHellIsThis** [AppMetrics/AppMetrics#261]
Ok, I'm a bit confused here.  I want to use AppMetrics with a WEBAPI project that is strictly a REST API no web pages, no Razor required etc.

The closest instructions seem to be these instructions: https://www.app-metrics.io/web-monitoring/aspnet-core/quick-start/

I installed the package `App.Metrics.AspNetCore.Mvc ` as instructed but it pulls in a couple of RAZOR packages automatically.

Is this required for a webapi only project?

**Comment by alhardy** [comment]
If you don't want a dependancy on 'Microsoft.AspNetCore.Mvc'  you can use  the 
[App.Metrics.AspNetCore](https://www.nuget.org/packages/App.Metrics.AspNetCore/) package

**Comment by WhatFreshHellIsThis** [comment]
Is the Microsoft.AspNetCore.Mvc package required to track metrics per endpoint automatically?
(Am I correct in assuming that if I figure this all out I should end up with metrics per API endpoint, i.e. if I have 50 endpoints I will end up with at least 50 metrics tracked?)

**Comment by alhardy** [comment]
Correct, unless they are excluded my the ignored route setting and provided you have attached the MVC route handler using `.AddMvc().AddMetrics()`. 

Without using MVC, endpoints with matching actions will be counted as the same e.g. `/users` and `/users/{I'd}`. 

See the [routing package](https://github.com/AppMetrics/AspNetCore/blob/dev/src/App.Metrics.AspNetCore.Routing/DefaultMetricsRouteNameResolver.cs)

**Comment by WhatFreshHellIsThis** [comment]
Hmm...I'm not seeing any of my endpoints being tracked individually, just a lot of GET and PUT and DELETE as if they are being munged together.  I have dozens of routes and am testing the metrics by running tests that exercise all the routes automatically but none of them show up individually by name in the metrics even though clearly a lot of data is being gathered.

**Comment by alhardy** [comment]
@WhatFreshHellIsThis Here's part of the [app metrics grafana dashboard](https://grafana.com/dashboards/2125) using the [influxdb sample](https://github.com/AppMetrics/InfluxDB/tree/master/sandbox/MetricsInfluxDBSandboxMvc) showing all endpoints tracked individually.

![image](https://user-images.githubusercontent.com/729185/39359821-e0bdeb92-4a5e-11e8-8b0a-f3042798e95c.png)

And a snippet of the `/metrics-text` endpoint in the same sample

![image](https://user-images.githubusercontent.com/729185/39359879-2739ac64-4a5f-11e8-8220-7dade504ad12.png)

If you can provide a sample in a github repo I'm happy to take a look at your configuration to help getting it to work?

**Comment by WhatFreshHellIsThis** [comment]
Thank you for including in the revised docs a reference to the method GetMetricsCurrentRouteName to check what Metrics *thinks* is the route name.  It revealed exactly the problem I'm seeing.

I'm wondering if I am missing a setting that would cause this behaviour or my routes are too complex for it as I'm versioning my routes and in some cases I need to define them a little differently to support other needs.

Here is what I'm seeing:

I've put this line of code into 3 test controller routes:
`var routeTemplate = HttpContext.GetMetricsCurrentRouteName();`

It's revealed that routeTemplate is always the string  "GET" for the routes below:

Here is the "Widget" test controller class definition:
```
 [ApiVersion("8.0")]
    [Route("api/v{version:apiVersion}/[controller]")]
    [Produces("application/json")]
    [Authorize]
    public class WidgetController : Controller
{...etc
```
Here are the definitions for the individual routes I just tested, in all 3 cases GetMetricsCurrentRouteName returns simply "GET":

```
 [HttpGet("TestWidgetJob")]
        public ActionResult TestWidgetJob()
        {
```

```
 [HttpGet("List", Name = nameof(List))]
        public async Task<IActionResult> List([FromQuery] PagingOptions pagingOptions)
        {
```

```
 [HttpGet("PickList", Name = nameof(WidgetPickList))]
        public async Task<IActionResult> WidgetPickList([FromQuery] string q, [FromQuery] PagingOptions pagingOptions)
        {
```

When I run integration tests across all the routes in my api the resulting metrics only show "GET", "PUT", etc, no route names, it's aggregating them all together.

Is there a missing or incorrect setting that could cause this behaviour?

**Comment by alhardy** [comment]
Thanks, this helps diagnose.

Could you also post the code in your `Startup.cs`, looking to see how api versioning has been configured.

**Comment by alhardy** [comment]
@WhatFreshHellIsThis  that's a big `Startup.cs`, 385 lines.

I've added a [sample app](https://github.com/AppMetrics/Samples.V2/blob/master/AspNetCore2.VersionedApi.QuickStart/Startup.cs#L18) that demo's using App Metrics which AspNet Core MVC versioning.

Appears to be tagging by route correctly.

Caveat is, if I request an "unsupported" version, I see metrics tagged with just a `GET` route tag as you do.

This is probably because App Metrics is unable to resolve a route due to it being "unsupported". I guess I could investigate if it's possible to retrieve the "unsupported" route somehow any idea's if that is possible?

E.g. in the sample app request `api/v1.0/values` and `api/v2.0/values`, this works find in terms of App Metrics tagging. Then request `api/1.2/values` which returns "UnSupportedApiVersion" and results in metrics tagged with route of `GET`. App Metrics is only attempting to record these metrics when configured to "track" 400 status codes.

**Comment by WhatFreshHellIsThis** [comment]
Ok, so you got way further than I have if you are getting metrics tagging with the versioned routes working at all.

I'll have to dig into your sample code because I'm not getting the same result at all.

I'm not having any issues with unsupported routes because I'm not testing with any, it's the **supported** versioned routes that are not working with App.Metrics properly.  

I don't understand what the issue you're seeing with unsupported routes is exactly (or why it's an issue), metrics for unsupported and unimplemented routes sounds fine as is, it *should* return an error code and it can't log it properly as it's "unsupported" just like any other "gibberish" request to the API server.  

My issue is with supported routes, I'll do some digging...

**Comment by WhatFreshHellIsThis** [comment]
Found it!  It was a mistake and should have jumped out at me much sooner as it's really obvious in hindsight, sorry about that @alhardy,  thank you for your help and I hope others in future can benefit from my mistake.

In Startup.cs this block of code works properly:
```
 services.AddMvc()
  .AddMetrics();
```

However this block of code results in no errors anywhere during compile or runtime but the routes are not identified and tagged and instead all routes only show up as the HTTP method (e.g. GET, PUT etc):

```
services.AddMvc();
services.AddMetrics();
```

**Comment by alhardy** [comment]
Yes, your first example is an extension on IMvcBuilder, the second is on IServiceCollection ;)

All good, ended up finding an interesting edge case out of this so was worth the investigation

**Comment by rdcm** [comment]
@alhardy Can i use `AddMetrics();` without `.AddMvc()` ? `.AddMvc()` method add redundant dependencies for web api app:

```
...
      builder.AddRazorViewEngine();
      builder.AddRazorPages();
      builder.AddCacheTagHelper();
...
```

In you sample `AddMvc()` and `AddMvcCore()` used together, but usually chosen one.

**Comment by alhardy** [comment]
@rdcm in the next version yes you'll be able to use either e.g. https://github.com/AppMetrics/AspNetCore/blob/dev/sandbox/MetricsSandboxMvc/Startup.cs#L32

**Comment by FastNinja** [comment]
Just experienced same issue. almost gave up. found the issue. Thanks @WhatFreshHellIsThis

**Comment by FastNinja** [comment]
@alhardy is there a timeline for when I can use `MvcCore` and `Metrics`
it looks like I can't do it right now
```.AddMvcCore.AddMetrics()```

and we are not planning to add full MVC for just API project.

Is there a workaround or anything I can do right now so Route endpoints will be captured?

Experiencing same issue where route is always just verb `route="GET"` and all routing information is lost, making metrics not that useful (so they are now only for the whole application only..)

**Comment by alhardy** [comment]
@FastNinja It's already released

**Comment by bardagadvergur** [comment]
@alhardy I am getting these tags (v{version:apiversion}) . Any idea why? I've looked through your sample but can't spot the difference:

TIMESTAMP: 636754755540559341
MEASUREMENT: [Application.HttpRequests] Transactions Per Endpoint
TAGS:
  route = GET api/v{version:apiversion}/items/info

**Comment by bardagadvergur** [comment]
Your test proj seems to do the same:
![2018-10-18_20h15_15](https://user-images.githubusercontent.com/16955820/47181552-c0a96880-d312-11e8-844b-b48d4d74a73d.png)

**Comment by alhardy** [comment]
Not sure if this has changed in a new version of asp.net core. Experience this issue too so need to have a look

**Comment by bardagadvergur** [comment]
@alhardy have you had a chance to look into this?

**Comment by alhardy** [comment]
No I haven't had a chance to look into the asp.net routing related changes yet
